### DISCUSSION TITLE
Transform texture scale does not work in IOS

### DISCUSSION THREAD

**Discussion by KESUKE22** [google/model-viewer#4403] [Q&A] [answered]
On the andorid device, rotation, offset, and scale all work, but on IOS, rotation and offset work, but when I adjust the scale and press the AR button, the button disappears and the session does not start!
The reproduction environment is confirmed by opening the following page with a smart phone

https://modelviewer.dev/examples/scenegraph/#transformTexturesExample

**Comment by elalish** [comment]
@mrdoob Have you tried USDZExporter with a texture scale applied?

**Reply by KESUKE22** [reply]
"Thank you, I believe I've received a very valuable hint, but I don't understand the meaning of 'USDZ Exporter with a tex~'. I would be happy if you could tell me how to try it."

**Reply by mrdoob** [reply]
Pretty sure @hybridherbst tested this 🤔

**Reply by hybridherbst** [reply]
Oh yes, certainly. That should work.

Quick glance at the console shows:
<img width="1012" alt="image" src="https://github.com/google/model-viewer/assets/2693840/2e3889c7-3de6-40c8-a3ac-4c13b754b0bc">
which would mean that the data passed to repeat isn't actually a Vector2?

**Reply by hybridherbst** [reply]
Which is most likely from here:
https://github.com/mrdoob/three.js/blob/1c2c80c2638f80accf329a47b2d47279f0362a14/examples/jsm/exporters/USDZExporter.js#L452C1-L454

As repeat is originally a Vector
https://github.com/mrdoob/three.js/blob/1c2c80c2638f80accf329a47b2d47279f0362a14/src/textures/Texture.js#L57
I'd say maybe model-viewer code doesn't set it correctly as a vector?

At least in the past my attempts at making three.js PRs that sanitize data on export were declined ;) passing correct data in is application scope.

**Comment by hybridherbst** [comment, marked as answer]
This is caused by the sample code itself:
`sampler.setScale({x: 0, y: 1})` does not match the `Vector2 | null` signature. Either the sample should be adjusted or the setScale method needs to handle that case.

https://github.com/google/model-viewer/blob/7aed15045790129e73bb56de9642430351c02a56/packages/model-viewer/src/features/scene-graph/sampler.ts#L148-L154

**Reply by elalish** [reply]
Thank you! @diegoteran would you mind taking a look?

**Reply by elalish** [reply]
Ah, I totally missed this in code review - model-viewer's public API should not reference any three.js types.

**Reply by hybridherbst** [reply]
I see! - in that case setScale et al would probably need to handle the conversion to a THREE.Vector2 internally.

**Reply by diegoteran** [reply]
Changes should be in #4473.

**Comment by echadwick-artist** [comment]
Another example https://ericchadwick.com/img/ecommerce.html#SheenChair

Windows 11 Home 22H2 22621.2134, Chrome 116.0.5845.97 (Official Build) (64-bit):
![Windows-desktop-Chrome](https://github.com/google/model-viewer/assets/40177036/211eeecb-c117-44cf-9484-8170a7edaa09)

iOS desktop Safari:
![iOS-desktop-Safari](https://github.com/google/model-viewer/assets/40177036/c8b6ece7-cfca-4ab3-ad52-c874bc3cf15c)

iOS desktop Chrome (and same for ):
![iOS-desktop-Chrome](https://github.com/google/model-viewer/assets/40177036/7c3e5b73-d167-43e7-8ea6-f0436a8f400c)

Also occurs in iOS desktop Firefox, and in Safari iPadOS 17beta (<model-viewer>, not ARQL).
